### PR TITLE
clp-package: Add support for running package as root (fixes #500).

### LINKED ISSUE #500: CLP package can't be started by root user
### Bug

When running as root, the clp-package fails to start. `sbin/start-clp.sh` prints the following log:

```
Traceback (most recent call last):
  File "/home/local/tmp/clp-text-ubuntu-focal-x86_64-v0.1.1/lib/python3/site-packages/clp_package_utils/scripts/start_clp.py", line 943, in main
    start_queue(instance_id, clp_config)
  File "/home/local/tmp/clp-text-ubuntu-focal-x86_64-v0.1.1/lib/python3/site-packages/clp_package_utils/scripts/start_clp.py", line 295, in start_queue
    raise EnvironmentError(f"{component_name} did not initialize in time")
OSError: queue did not initialize in time
```

Checking the queue container's logs shows:

```
2024-07-29 04:11:47.440453+00:00 [erro] <0.145.0> Error when reading /var/lib/rabbitmq/.erlang.cookie: eacces
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>     supervisor: {local,net_sup}
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>     errorContext: start_error
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>     reason: {"Error when reading /var/lib/rabbitmq/.erlang.cookie: eacces",
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>              [{auth,init_no_setcookie,0,[{file,"auth.erl"},{line,293}]},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>               {auth,init,1,[{file,"auth.erl"},{line,144}]},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>               {gen_server,init_it,2,[{file,"gen_server.erl"},{line,423}]},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>               {gen_server,init_it,6,[{file,"gen_server.erl"},{line,390}]},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>               {proc_lib,init_p_do_apply,3,
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>                         [{file,"proc_lib.erl"},{line,226}]}]}
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>     offender: [{pid,undefined},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>                {id,auth},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>                {mfargs,{auth,start_link,[]}},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>                {restart_type,permanent},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>                {significant,false},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>                {shutdown,2000},
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>                {child_type,worker}]
2024-07-29 04:11:47.453762+00:00 [erro] <0.143.0>
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>   crasher:
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     initial call: auth:init/1
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     pid: <0.145.0>
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     registered_name: []
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     exception error: "Error when reading /var/lib/rabbitmq/.erlang.cookie: eacces"
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>       in function  auth:init_no_setcookie/0 (auth.erl, line 293)
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>       in call from auth:init/1 (auth.erl, line 144)
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>       in call from gen_server:init_it/2 (gen_server.erl, line 423)
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>       in call from gen_server:init_it/6 (gen_server.erl, line 390)
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     ancestors: [net_sup,kernel_sup,<0.47.0>]
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     message_queue_len: 0
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     messages: []
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     links: [<0.143.0>]
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     dictionary: []
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     trap_exit: true
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     status: running
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     heap_size: 987
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     stack_size: 29
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>     reductions: 2078
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>   neighbours:
2024-07-29 04:11:47.453885+00:00 [erro] <0.145.0>

BOOT FAILED
...
```

A similar issue was reported in #135.

### CLP version

0.1.1

### Environment

* OS: Ubuntu 22.04
* Docker: 24.0.6

### Reproduction steps

As root, run:

```shell
curl -fsSLO https://github.com/y-scope/clp/releases/download/v0.1.1/clp-text-ubuntu-focal-x86_64-v0.1.1.tar.gz
tar xzf clp-text-ubuntu-focal-x86_64-v0.1.1.tar.gz
cd clp-text-ubuntu-focal-x86_64-v0.1.1
sbin/start-clp.sh
```

### PR DESCRIPTION
**Pull request by haiqi96**
# Description
When running container as root user, redis, results cache and rabbitmq containers lose the write permission to log and data directories.
This PR adds a temporary fix. We let the container run with their default user and change the ownership of the log directories when the current user is root.
Also adds a start up check for redis

# Validation performed
1. Run clp package with root, confirmed that package can run properly
2. Run clp package with non-root. Stop and restart the same package as root, confirmed that package can run properly, indicating that ownership of the folders and their contents has been changed successfully

### CODE REVIEW

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py [hunk 1]
```diff
@@ -268,6 +277,14 @@ def start_queue(instance_id: str, clp_config: CLPConfig):
         DockerMount(DockerMountType.BIND, queue_logs_dir, rabbitmq_logs_dir),
     ]
     rabbitmq_pid_file_path = pathlib.Path("/") / "tmp" / "rabbitmq.pid"
+
+    container_user = f"{os.getuid()}:{os.getgid()}"
```
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:281]: We can factor this code into a function as well:
```
def set_permission_and_get_user(list_paths: List[str]):
    container_user = f"{os.getuid()}:{os.getgid()}"
    if os.getuid() == 0:
        # Use the default container user and make the directories
        # writable by the default user if the current user is root
        container_user = "mongodb"
        for path in list_paths:
            chown_recursively(path, 999, 999)
    return container_user
```

**Review by kirkrodrigues — DISMISSED** [review]
For the PR title, how about:

clp-package: Add support for running package as root (fixes #500).

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py [hunk 2]
```diff
@@ -268,6 +283,17 @@ def start_queue(instance_id: str, clp_config: CLPConfig):
...
+
+    host_user_id = os.getuid()
+    if 0 != host_user_id:
+        container_user = f"{host_user_id}:{os.getgid()}"
+    else:
+        # The host user is `root` so use the container's default user and make this component's
+        # directories writable by that user.
+        # NOTE: This doesn't affect the host user's access to the directories since they're `root`.
+        container_user = "rabbitmq"
+        chown_recursively(queue_logs_dir, 999, 999)
```
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:295]:
```suggestion
        default_container_user_id = 999
        default_container_group_id = 999
        chown_recursively(queue_logs_dir, default_container_user_id, default_container_group_id)
```

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py [hunk 3]
```diff
@@ -330,13 +356,27 @@ def start_redis(instance_id: str, clp_config: CLPConfig, conf_dir: pathlib.Path)
...
+        container_user = f"{host_user_id}:{os.getgid()}"
+    else:
+        # The host user is `root` so use the container's default user and make this component's
+        # directories writable by that user.
+        # NOTE: This doesn't affect the host user's access to the directories since they're `root`.
+        container_user = "redis"
+        container_user_id = 999
+        container_group_id = 999
+        chown_recursively(redis_data_dir, container_user_id, container_group_id)
+        chown_recursively(redis_logs_dir, container_user_id, container_group_id)
```
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:371]:
```suggestion
        default_container_user_id = 999
        default_container_group_id = 999
        chown_recursively(redis_data_dir, default_container_user_id, default_container_group_id)
        chown_recursively(redis_logs_dir, default_container_user_id, default_container_group_id)
```

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py [hunk 4]
```diff
@@ -376,13 +429,27 @@ def start_results_cache(instance_id: str, clp_config: CLPConfig, conf_dir: pathl
...
+        container_user = f"{host_user_id}:{os.getgid()}"
+    else:
+        # The host user is `root` so use the container's default user and make this component's
+        # directories writable by that user.
+        # NOTE: This doesn't affect the host user's access to the directories since they're `root`.
+        container_user = "mongodb"
+        container_user_id = 999
+        container_group_id = 999
+        chown_recursively(data_dir, container_user_id, container_group_id)
+        chown_recursively(logs_dir, container_user_id, container_group_id)
```
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:444]:
```suggestion
        default_container_user_id = 999
        default_container_group_id = 999
        chown_recursively(data_dir, default_container_user_id, default_container_group_id)
        chown_recursively(logs_dir, default_container_user_id, default_container_group_id)
```

**Review by kirkrodrigues — APPROVED** [review]

**Review by davidlion — APPROVED** [review]

**Pull request merged**